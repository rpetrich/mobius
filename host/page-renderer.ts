import { BootstrapData } from "internal-impl";
import { Root as CSSRoot } from "postcss";
import { once } from "./memoize";
import { newDocument, serialize } from "./redom";

function compatibleStringify(value: any): string {
	return JSON.stringify(value).replace(/\u2028/g, "\\u2028").replace(/\u2029/g, "\\u2029").replace(/<\/script/g, "<\\/script");
}

function migrateChildren(fromNode: Node, toNode: Node) {
	let firstChild: Node | undefined;
	while (firstChild = fromNode.firstChild) {
		toNode.appendChild(firstChild);
	}
}

export const enum PageRenderMode {
	Bare = 0,
	IncludeForm = 1,
	IncludeFormAndStripScript = 2,
}

export interface SessionState {
	readonly sessionID: string;
	readonly localChannelCount: number;
}

export interface ClientState {
	readonly clientID: number;
	incomingMessageId: number;
}

export interface SharedRenderState {
	readonly document: ReturnType<typeof newDocument>;
	readonly noscript: Element;
	readonly metaRedirect: Element;
	cssForPath(path: string): Promise<CSSRoot>;
}

export interface RenderOptions {
	mode: PageRenderMode;
	clientState: ClientState;
	sessionState: SessionState;
	clientURL: string;
	clientIntegrity: string;
	fallbackIntegrity: string;
	fallbackURL: string;
	noScriptURL?: string;
	bootstrapData?: BootstrapData;
	inlineCSS?: true;
}

const cssRoot = once(async () => (await import("postcss")).root);

export function defaultDocument() {
	const document = newDocument();
	const noscript = document.createElement("noscript");
	const viewport = document.createElement("meta");
	viewport.setAttribute("name", "viewport");
	viewport.setAttribute("content", "width=device-width, initial-scale=1");
	document.head.appendChild(viewport);
	const dispatchScript = document.createElement("script");
	dispatchScript.textContent = `(_dispatch=function(i,e){_dispatch.e.push([i,e])}).e=[]`;
	document.head.appendChild(dispatchScript);
	document.body.setAttribute("data-gramm", "false");
	document.body.className = "notranslate";
	document.body.appendChild(document.createElement("div"));
	const metaRedirect = document.createElement("meta");
	metaRedirect.setAttribute("http-equiv", "refresh");
	noscript.appendChild(metaRedirect);
	return {
		document,
		noscript,
		metaRedirect,
	};
}

export class PageRenderer {
	public readonly document: ReturnType<typeof newDocument>;
	public readonly body: Element;
	public readonly head: Element;
	private clientScript: Element;
	private fallbackScript: Element;
	private inlineStyles?: Element;
	private bootstrapScript?: Element;
	private formNode?: Element;
	private postbackInput?: Element;
	private sessionIdInput?: Element;
	private clientIdInput?: Element;
	private messageIdInput?: Element;
	private hasServerChannelsInput?: Element;

	constructor(private sharedState: SharedRenderState) {
		this.document = sharedState.document.cloneNode(true);
		this.body = this.document.body;
		this.head = this.document.head;
		this.body.appendChild(this.clientScript = this.document.createElement("script"));
		this.body.appendChild(this.fallbackScript = this.document.createElement("script"));
	}

	// Render document state into an HTML document containing the appropriate bootstrap data and configuration
	public async render({ mode, clientState, sessionState, clientURL, clientIntegrity, fallbackIntegrity, fallbackURL, noScriptURL, bootstrapData, inlineCSS }: RenderOptions): Promise<string> {
		const document = this.document;
		let bootstrapScript: Element | undefined;
		let textNode: Node | undefined;
		let formNode: Element | undefined;
		let postbackInput: Element | undefined;
		let sessionIdInput: Element | undefined;
		let clientIdInput: Element | undefined;
		let messageIdInput: Element | undefined;
		let hasServerChannelsInput: Element | undefined;
		let siblingNode: Node | null = null;
		let cssRoots: Array<Promise<CSSRoot>> | undefined;
		// CSS Inlining
		if (inlineCSS) {
			const linkTags = document.getElementsByTagName("link");
			for (let i = 0; i < linkTags.length; i++) {
				if (linkTags[i].getAttribute("rel") === "stylesheet") {
					const href = linkTags[i].getAttribute("href");
					if (href && !/^\w+:/.test(href)) {
						const root = this.sharedState.cssForPath(href);
						if (cssRoots) {
							cssRoots.push(root);
						} else {
							cssRoots = [root];
						}
					}
				}
			}
		}
		// Hidden form elements for fallbacks
		if (mode >= PageRenderMode.IncludeForm) {
			formNode = this.formNode;
			if (!formNode) {
				formNode = this.formNode = document.createElement("form");
				formNode.setAttribute("action", "/");
				formNode.setAttribute("method", "POST");
				formNode.setAttribute("id", "mobius-form");
			}
			postbackInput = this.postbackInput;
			if (!postbackInput) {
				postbackInput = this.postbackInput = document.createElement("input");
				postbackInput.setAttribute("name", "postback");
				postbackInput.setAttribute("type", "hidden");
				postbackInput.setAttribute("value", "form");
			}
			formNode.appendChild(postbackInput);
			sessionIdInput = this.sessionIdInput;
			if (!sessionIdInput) {
				sessionIdInput = this.sessionIdInput = document.createElement("input");
				sessionIdInput.setAttribute("name", "sessionID");
				sessionIdInput.setAttribute("type", "hidden");
				sessionIdInput.setAttribute("value", sessionState.sessionID);
			}
			formNode.appendChild(sessionIdInput);
			if (clientState.clientID != 0) {
				clientIdInput = this.clientIdInput;
				if (!clientIdInput) {
					clientIdInput = this.clientIdInput = document.createElement("input");
					clientIdInput.setAttribute("name", "clientID");
					clientIdInput.setAttribute("type", "hidden");
				}
				clientIdInput.setAttribute("value", clientState.clientID.toString());
				formNode.appendChild(clientIdInput);
			}
			messageIdInput = this.messageIdInput;
			if (!messageIdInput) {
				messageIdInput = this.messageIdInput = document.createElement("input");
				messageIdInput.setAttribute("name", "messageID");
				messageIdInput.setAttribute("type", "hidden");
			}
			messageIdInput.setAttribute("value", clientState.incomingMessageId.toString());
			formNode.appendChild(messageIdInput);
			hasServerChannelsInput = this.hasServerChannelsInput;
			if (!hasServerChannelsInput) {
				hasServerChannelsInput = this.hasServerChannelsInput = document.createElement("input");
				hasServerChannelsInput.setAttribute("name", "hasServerChannels");
				hasServerChannelsInput.setAttribute("type", "hidden");
			}
			hasServerChannelsInput.setAttribute("value", sessionState.localChannelCount ? "1" : "");
			formNode.appendChild(hasServerChannelsInput);
			migrateChildren(this.body, formNode);
			this.body.appendChild(formNode);
		}
		if (mode >= PageRenderMode.IncludeFormAndStripScript) {
			siblingNode = document.createTextNode("");
			this.clientScript.parentNode!.insertBefore(siblingNode, this.clientScript);
			this.clientScript.parentNode!.removeChild(this.clientScript);
			this.fallbackScript.parentNode!.removeChild(this.fallbackScript);
		} else if (bootstrapData) {
			bootstrapScript = this.bootstrapScript;
			if (!bootstrapScript) {
				bootstrapScript = this.bootstrapScript = document.createElement("script");
				bootstrapScript.setAttribute("type", "application/x-mobius-bootstrap");
			}
			textNode = document.createTextNode(compatibleStringify(bootstrapData));
			bootstrapScript.appendChild(textNode);
			this.clientScript.parentNode!.insertBefore(bootstrapScript, this.clientScript);
		}
		this.clientScript.setAttribute("src", clientURL);
		this.clientScript.setAttribute("integrity", clientIntegrity);
		this.fallbackScript.textContent = `window._mobius||(function(s){s.src=${JSON.stringify(fallbackURL)};s.setAttribute("integrity",${JSON.stringify(fallbackIntegrity)})})(document.head.appendChild(document.createElement("script")))`;
		if (noScriptURL) {
			this.sharedState.metaRedirect.setAttribute("content", "0; url=" + noScriptURL);
			this.head.appendChild(this.sharedState.noscript);
		}
		try {
			if (cssRoots) {
				const newRoot = (await cssRoot())();
				for (const root of await Promise.all(cssRoots)) {
					if (root.nodes) {
						for (const node of root.nodes) {
							if (node.type === "rule") {
								try {
									if (document.querySelector(node.selector) === null) {
										continue;
									}
								} catch (e) {
									continue;
								}
								newRoot.append(node.clone());
							}
						}
					}
				}
				if (newRoot.nodes && newRoot.nodes.length) {
					let inlineStyles = this.inlineStyles;
					if (!inlineStyles) {
						inlineStyles = document.createElement("style");
						inlineStyles.setAttribute("id", "mobius-inlined");
						this.head.appendChild(inlineStyles);
						this.inlineStyles = inlineStyles;
					}
					inlineStyles.textContent = newRoot.toResult().css;
				}
			}
			return "<!doctype html>" + serialize(this.document, mode >= PageRenderMode.IncludeFormAndStripScript ? filterEventHandlers : undefined);
		} finally {
			// Put Humpty Dumpty back together again
			if (mode >= PageRenderMode.IncludeForm && formNode) {
				if (postbackInput) {
					formNode.removeChild(postbackInput);
				}
				if (sessionIdInput) {
					formNode.removeChild(sessionIdInput);
				}
				if (clientIdInput) {
					formNode.removeChild(clientIdInput);
				}
				if (messageIdInput) {
					formNode.removeChild(messageIdInput);
				}
				if (hasServerChannelsInput) {
					formNode.removeChild(hasServerChannelsInput);
				}
				migrateChildren(formNode, this.body);
				this.body.removeChild(formNode);
			}
			if (mode >= PageRenderMode.IncludeFormAndStripScript) {
				if (siblingNode) {
					siblingNode.parentNode!.insertBefore(this.fallbackScript, siblingNode);
					siblingNode.parentNode!.insertBefore(this.clientScript, siblingNode);
					siblingNode.parentNode!.removeChild(siblingNode);
				}
			}
			if (noScriptURL) {
				this.head.removeChild(this.sharedState.noscript);
			}
			if (bootstrapScript) {
				const parentElement = bootstrapScript.parentNode;
				if (parentElement) {
					parentElement.removeChild(bootstrapScript);
				}
				if (textNode) {
					bootstrapScript.removeChild(textNode);
				}
			}
		}
	}
}

function filterEventHandlers(attribute: Attribute) {
		return !/^on/.test(attribute.name);
}
