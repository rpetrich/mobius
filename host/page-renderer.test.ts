import { defaultDocument, PageRenderer, PageRenderMode } from "./page-renderer";

const defaults = {
	clientState: {
		clientID: 0,
		incomingMessageId: 0,
	},
	sessionState: {
		sessionID: "test",
		localChannelCount: 0,
	},
	clientURL: "/main.js",
	clientIntegrity: "sha1-main-dummy",
	fallbackURL: "/fallback.js",
	fallbackIntegrity: "sha1-fallback-dummy",
};

describe("PageRenderer", () => {
	const renderer = new PageRenderer({
		async cssForPath() {
			throw new Error("No CSS!");
		},
		...defaultDocument(),
	});
	it("bare render", async () => {
		expect(await renderer.render({
			mode: PageRenderMode.Bare,
			...defaults,
		})).toMatchSnapshot();
	});
	it("with form", async () => {
		expect(await renderer.render({
			mode: PageRenderMode.IncludeForm,
			...defaults,
		})).toMatchSnapshot();
	});
	it("with form and strip script", async () => {
		expect(await renderer.render({
			mode: PageRenderMode.IncludeFormAndStripScript,
			...defaults,
		})).toMatchSnapshot();
	});
});
