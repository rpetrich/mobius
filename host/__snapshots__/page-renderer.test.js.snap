// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PageRenderer bare render 1`] = `"<!doctype html><html><head><meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1\\"><script>(_dispatch=function(i,e){_dispatch.e.push([i,e])}).e=[]</script></head><body data-gramm=\\"false\\" class=\\"notranslate\\"><div></div><script src=\\"/main.js\\" integrity=\\"sha1-main-dummy\\"></script><script>window._mobius||(function(s){s.src=\\"/fallback.js\\";s.setAttribute(\\"integrity\\",\\"sha1-fallback-dummy\\")})(document.head.appendChild(document.createElement(\\"script\\")))</script></body></html>"`;

exports[`PageRenderer with form 1`] = `"<!doctype html><html><head><meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1\\"><script>(_dispatch=function(i,e){_dispatch.e.push([i,e])}).e=[]</script></head><body data-gramm=\\"false\\" class=\\"notranslate\\"><form action=\\"/\\" method=\\"POST\\" id=\\"mobius-form\\"><input name=\\"postback\\" type=\\"hidden\\" value=\\"form\\"><input name=\\"sessionID\\" type=\\"hidden\\" value=\\"test\\"><input name=\\"messageID\\" type=\\"hidden\\" value=\\"0\\"><input name=\\"hasServerChannels\\" type=\\"hidden\\" value><div></div><script src=\\"/main.js\\" integrity=\\"sha1-main-dummy\\"></script><script>window._mobius||(function(s){s.src=\\"/fallback.js\\";s.setAttribute(\\"integrity\\",\\"sha1-fallback-dummy\\")})(document.head.appendChild(document.createElement(\\"script\\")))</script></form></body></html>"`;

exports[`PageRenderer with form and strip script 1`] = `"<!doctype html><html><head><meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1\\"><script>(_dispatch=function(i,e){_dispatch.e.push([i,e])}).e=[]</script></head><body data-gramm=\\"false\\" class=\\"notranslate\\"><form action=\\"/\\" method=\\"POST\\" id=\\"mobius-form\\"><input name=\\"postback\\" type=\\"hidden\\" value=\\"form\\"><input name=\\"sessionID\\" type=\\"hidden\\" value=\\"test\\"><input name=\\"messageID\\" type=\\"hidden\\" value=\\"0\\"><input name=\\"hasServerChannels\\" type=\\"hidden\\" value><div></div></form></body></html>"`;
